### PR TITLE
Changes for Telemetry in Azure Test Plan Task

### PR DESCRIPTION
**Pull request by adityashahms**
**Task name**:  AzureTestPlanV0

**Description**: Adding CI telemetry to drive usage metrics

**Documentation changes required:** (Y/N) N

**Added unit tests:** (Y/N) N

**Attached related issue:** (Y/N) N

**Checklist**:
- [x] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md) how to do it
- [x] Checked that applied changes work as expected

**Testing**:
- Sample JSON Body Object for Maven:
```
{
  "builduri": "vstfs:///Build/Build/553",
  "buildid": "553",
  "TestSelector": "automatedTests,manualTests",
  "totalNumOfManualTestPoint": 3,
  "totalNumOfAutomatedTestPoint": 4,
  "totalNumOfTestSuites": 1,
  "manualTestFlowReturnCode": 0,
  "isJavaMavenExecution": true,    // Incase of multi-language. We would have the rest of the language objects as well
  "automatedTestFlowReturnCode": 1
}

```
- Pipeline Run
image
- Run Information and Result Breakdown
This information is available from Test Run Feature Telemetry and PublishTestResult Telemetry since we have leveraged APIs and PTR library directly.

### CODE REVIEW

File: Tasks/AzureTestPlanV0/ciEventLogger.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import * as tl from 'azure-pipelines-task-lib/task';
+
+const area: string = 'TestManagement';
+const feature: string = 'AzureTestPlanTask';
+
+function getDefaultProps() {
+    return {
+        releaseuri: tl.getVariable('Release.ReleaseUri'),
```
Comment (eswarnathp) [inline Tasks/AzureTestPlanV0/ciEventLogger.ts:8]: can we check if it is release / build flow and return only those params?

File: Tasks/AzureTestPlanV0/ciEventLogger.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    };
+}
+
+export function publishEvent(properties: { [key: string]: any }): void {
+    try {
+        tl.assertAgent('2.144.0');
+        publishTelemetry(area, feature, Object.assign(getDefaultProps(), properties));
+
+    } catch (err) {
+        tl.debug('Unable to publish telemetry due to lower agent version.');
```
Comment (eswarnathp) [inline Tasks/AzureTestPlanV0/ciEventLogger.ts:21]: will the error message be same in all the cases?

File: Tasks/AzureTestPlanV0/ciEventLogger.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    }
+}
+
+export function publishTelemetry(area: string, feature: string, properties: { [key: string]: any }): void {
+    const data = JSON.stringify(properties);
+    tl.debug('telemetry area: ' + area + ' feature: ' + feature + ' data: ' + data);
+    tl.command('telemetry.publish', { 'area': area, 'feature': feature }, data);
```
Comment (eswarnathp) [inline Tasks/AzureTestPlanV0/ciEventLogger.ts:28]: should we log all the telemetry in the logs? I think its not required, you can remove data

**Review by eswarnathp — COMMENTED** [review]
Added comments

File: Tasks/AzureTestPlanV0/ciEventLogger.ts (hunk 1 above)
Comment (adityashahms) [reply Tasks/AzureTestPlanV0/ciEventLogger.ts:8]: If it is build only scenario, releaseuri and releaseid would be null and wouldn't be populated in ciData.

File: Tasks/AzureTestPlanV0/ciEventLogger.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        releaseuri: tl.getVariable('Release.ReleaseUri'),
+        releaseid: tl.getVariable('Release.ReleaseId'),
+        builduri: tl.getVariable('Build.BuildUri'),
+        buildid: tl.getVariable('Build.Buildid')
+    };
+}
+
+export function publishEvent(properties: { [key: string]: any }): void {
+    try {
+        tl.assertAgent('2.144.0');
```
[the remark this answers is not in the document]
Comment (adityashahms) [reply Tasks/AzureTestPlanV0/ciEventLogger.ts:17]: Yes. This is hard dependency for the support of the telemetry.publish command.

File: Tasks/AzureTestPlanV0/ciEventLogger.ts (hunk 2 above)
Comment (adityashahms) [reply Tasks/AzureTestPlanV0/ciEventLogger.ts:21]: publishTelemetry is simply invoking async command via cmdlet and also not waiting for it(This is the way we have been using telemetry publish command). The only possible source of error can be the assert on agent version.

File: Tasks/AzureTestPlanV0/ciEventLogger.ts (hunk 3 above)
Comment (adityashahms) [reply Tasks/AzureTestPlanV0/ciEventLogger.ts:28]: I picked from VsTestV2, incase of telemtry.publish cmdlet failure, you have a fallback to look for telemetry in debug logs. 

We wouldn't log all the telemetry unless debug mode is enabled.

**Review by eswarnathp — APPROVED** [review]
lgtm

**Review by triptijain2112 — APPROVED** [review]

**Pull request merged**